controllers/helmchartinstall_controller.go

```diff
@@ -77,61 +126,255 @@ func (r *HelmChartInstallReconciler) render(ctx context.Context, claim *helmv1al
+		r.Log.V(0).Info("Couldn't find resources for configured behavior!", "claim", claim, "configuration", configuration)
+		return nil, nil
+
+	}
+
+	return &targetResourceBehavior, nil
+}
+
+/**
+ * When a behavior executes, the resource engine is configured by the
+ * object which triggered the behavior. This method encapsulates the logic to
+ * create the resource engine configuration from the object's fields.
+ *
+ * TODO Currently creation fails if the config map already exists, but it should
+ * succeed instead.
+ */
+func (r *HelmChartInstallReconciler) createBehaviorEngineConfiguration(
+	ctx context.Context,
+	claim *unstructured.Unstructured,
+	stackConfig *v1alpha1.StackConfiguration,
+) (*corev1.ConfigMap, error) {
+	// yamlyamlyamlyamlyaml
+	// TODO if spec is missing, that won't work very well
+	spec, ok := claim.Object["spec"]
+
+	if !ok {
+		r.Log.V(0).Info("Spec not found on claim; not creating engine configuration", "claim", claim)
+	}
+
+	r.Log.V(0).Info("Converting configuration", "spec", spec)
+	configContents, err := yaml.Marshal(spec)
+
+	r.Log.V(0).Info("Configuration contents as yaml", "configContents", configContents)
+
+	if err != nil {
+		r.Log.Error(err, "Error marshaling claim spec as yaml!", "claim", claim)
+		return nil, err
+	}
+
+	// Underneath, the yamler uses https://godoc.org/encoding/json#Marshal,
+	// which means that the bytes are UTF-8 encoded
+	// Theoretically we could get better performance by using a binary config
+	// map, but having a string makes it better for humans who may want to observe
+	// or troubleshoot behavior.
+	stringConfigContents := string(configContents)
+
+	// TODO get the engine type from the configuration
+	engineType := r.getEngineType(claim, stackConfig)
+
+	// TODO engine type should have a bit more structure;
+	// probably better to use an enum type pattern, with an
+	// engine name and its corresponding configuration file
+	// name in the same object
+	configKeyName := ""
+
+	if engineType == "helm2" {
+		configKeyName = "values.yaml"
+	}
+
+	configName := string(claim.GetUID())
+	generatedMap, err := r.generateConfigMap(configName, configKeyName, stringConfigContents)
+
+	if err != nil {
+		r.Log.V(0).Info("Error generating config map!", "claim", claim, "error", err)
+		return nil, err
+	}
+
+	generatedMap.SetNamespace(claim.GetNamespace())
+
+	r.Log.V(0).Info("Generated config map to pass engine configuration", "configMap", generatedMap)
+
+	err = r.Client.Create(ctx, generatedMap)
+
+	if err != nil {
+		r.Log.V(0).Info("Error creating config map!", "claim", claim, "error", err, "configMap", generatedMap)
+		return nil, err
+	}
+
+	return generatedMap, err
+}
+
+// The main reason this exists as its own method is to encapsulate the hashing logic
+func (r *HelmChartInstallReconciler) generateConfigMap(name string, fileName string, fileContents string) (*corev1.ConfigMap, error) {
+	configMap := &corev1.ConfigMap{}
```

Comment:
I would probably define `configMap` with the next 4 lines as one statement.

I'm curious about how and why this is being used.  I noticed that `hash.ConfigMapHash` takes the name into account, so it is interesting that the name is then changed, which invalidates the hash unless you know to remove the hash from the key.  Is this function just being used as a stand-in for a UID?

Reply:
> Is this function just being used as a stand-in for a UID?

Partially. The `UID`-like aspect is that it is used to power a pattern where config maps are treated as immutable. The hashing part gives it deterministic uniqueness based on the contents, which is helpful for deduping. It's not intended to be used for signature verification or checksumming or that type of thing (as far as I know).

The inspiration for this comes from the [kubectl functionality for creating config maps](https://github.com/kubernetes/kubectl/blob/8ebe75d8ba577459aa7377647e9a1d32eb4179a5/pkg/generate/versioned/configmap.go#L128).